### PR TITLE
[RFC] Non google-test Unit Tests Must Die + MSVC 2017 compliance

### LINKED ISSUE #62: Finalize switch to GTest unit-tests
Development has been started in #50

### LINKED ISSUE #213: Add sequential forallN tests
`forallN` is currently only tested when OpenMP is enabled.

### PR DESCRIPTION
**Pull request by willkill07**
Closes #62
Closes #213 

I updated the scan tests (they compile over 3x faster) and added two new checks `exclusive_offset` and `exclusive_inplace_offset`.

I added macros for using CUDA within google-test: `CUDA_TYPED_TEST_P`, `CUDA_TEST_P`, `CUDA_TEST_F`, and `CUDA_TEST` live under `test/include/RAJA_gtest.h`

This also updates all tests to run and pass on AppVeyor, so we can "claim" MSVC compliance with our test suite.

### CODE REVIEW

**Comment by willkill07** [comment]
I did make one change which could be considered breaking: raja_add_test automatically pulls in the required googletest object files, reducing the amount of CMake cruft required when specifying a test.

I also renamed gpu -> coda for the CUDA unit tests

**Comment by willkill07** [comment]
Also, If I understand the BLT PR, we can fully remove the thirdparty CMake config (making it much easier to digest IMO)

File: test/unit/cuda/test-reduce-max.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  }
+  return dvalue;
+}
+
+//
+// test 1 runs 3 reductions over a range multiple times to check
+//        that reduction value can be retrieved and then subsequent
+//        reductions can be run with the same reduction objects.
+//        Also exercises the get function call
+void test1()
```
Comment (willkill07) [inline test/unit/cuda/test-reduce-max.cpp]: I didn't really **want** to do it this way, but gtest does the weird class scoping which breaks NVCC.

Comment (davidbeckingsale) [inline test/unit/cuda/test-reduce-max.cpp]: How about:
```
#define CUDA_TEST(X, Y) \
static void cuda_test_ ## X ## Y();\
TEST(X,Y) { cuda_test_ ## X ## Y();}\
static void cuda_test_ ## X ## Y()
```

then use `CUDA_TEST(ReduceMax, Generic)`

Comment (davidbeckingsale) [inline test/unit/cuda/test-reduce-max.cpp]: We have used that in other places, and it hides the weirdness.

Comment (willkill07) [inline test/unit/cuda/test-reduce-max.cpp]: oh, good call here. Let me make the rounds

**Comment by willkill07** [comment]
I'd like to hold off on merging this for awhile -- I added CUDA_TEST, CUDA_TYPED_TEST_P and CUDA_TEST_F macros, so I need to revisit the other tests and update accordingly

**Comment by davidbeckingsale** [comment]
Sure thing, adding the extra macros is 👌

**Comment by willkill07** [comment]
Alright. Once Travis clears, this has my blessing to be merged. I updated the scan tests (they compile about 30% faster) and added two new checks `exclusive_offset` and `exclusive_inplace_offset`.

I'm using `CUDA_TYPED_TEST_P`, `CUDA_TEST_F`, and `CUDA_TEST` wherever appropriate

**Comment by willkill07** [comment]
Closes #213

**Comment by willkill07** [comment]
Idea for CUDA_TEST_P:

Here's the original TEST_P:
```cpp
# define TEST_P(test_case_name, test_name) \
  class GTEST_TEST_CLASS_NAME_(test_case_name, test_name) \
      : public test_case_name { \
   public: \
    GTEST_TEST_CLASS_NAME_(test_case_name, test_name)() {} \
    virtual void TestBody(); \
   private: \
    static int AddToRegistry() { \
      ::testing::UnitTest::GetInstance()->parameterized_test_registry(). \
          GetTestCasePatternHolder<test_case_name>(\
              #test_case_name, \
              ::testing::internal::CodeLocation(\
                  __FILE__, __LINE__))->AddTestPattern(\
                      #test_case_name, \
                      #test_name, \
                      new ::testing::internal::TestMetaFactory< \
                          GTEST_TEST_CLASS_NAME_(\
                              test_case_name, test_name)>()); \
      return 0; \
    } \
    static int gtest_registering_dummy_ GTEST_ATTRIBUTE_UNUSED_; \
    GTEST_DISALLOW_COPY_AND_ASSIGN_(\
        GTEST_TEST_CLASS_NAME_(test_case_name, test_name)); \
  }; \
  int GTEST_TEST_CLASS_NAME_(test_case_name, \
                             test_name)::gtest_registering_dummy_ = \
      GTEST_TEST_CLASS_NAME_(test_case_name, test_name)::AddToRegistry(); \
  void GTEST_TEST_CLASS_NAME_(test_case_name, test_name)::TestBody()
```

Proposed:
```cpp
# define TEST_P(test_case_name, test_name) \
  template <typename Invocable> \
  static void gtest_cuda_##test_case_name##_##test_name(Invocable &&); \
  class GTEST_TEST_CLASS_NAME_(test_case_name, test_name) \
      : public test_case_name { \
   public: \
    GTEST_TEST_CLASS_NAME_(test_case_name, test_name)() {} \
    virtual void TestBody() { \
      gtest_cuda_##test_case_name##_##test_name ([&] { return GetParam(); }); \
    }\
   private: \
    static int AddToRegistry() { \
      ::testing::UnitTest::GetInstance()->parameterized_test_registry(). \
          GetTestCasePatternHolder<test_case_name>(\
              #test_case_name, \
              ::testing::internal::CodeLocation(\
                  __FILE__, __LINE__))->AddTestPattern(\
                      #test_case_name, \
                      #test_name, \
                      new ::testing::internal::TestMetaFactory< \
                          GTEST_TEST_CLASS_NAME_(\
                              test_case_name, test_name)>()); \
      return 0; \
    } \
    static int gtest_registering_dummy_ GTEST_ATTRIBUTE_UNUSED_; \
    GTEST_DISALLOW_COPY_AND_ASSIGN_(\
        GTEST_TEST_CLASS_NAME_(test_case_name, test_name)); \
  }; \
  int GTEST_TEST_CLASS_NAME_(test_case_name, \
                             test_name)::gtest_registering_dummy_ = \
      GTEST_TEST_CLASS_NAME_(test_case_name, test_name)::AddToRegistry(); \
  template <typename Invocable> \
  static void gtest_cuda_##test_case_name##_##test_name(Invocable && GetParam)
```

TL;DR -- declare TestBody() to call the free function generated. The parameter to this free function is named GetParam and is a invocable type that curry's the tests's `GetParam()`. The usage would be the same.

**Comment by davidbeckingsale** [comment]
I joked about this in the meeting, but can you add a little bit of documentation/comment about the new macros for the benefit of people adding new tests?

**Comment by willkill07** [comment]
@davidbeckingsale Comments added to each of the macros. Additionally, per a recommendation from @trws, we now generate a single binary for unit testing.

I tried to do it "The CMake Way" but the CUDA specialization makes `target_sources()` break in horrible ways. I've added two macros to `test/CMakeLists.txt` that aid in visitation of the test subdirectory.

This should be good to go. 👍

**Comment by davidbeckingsale** [comment]
What's the advantage of the single binary?

**Comment by trws** [comment]
Build and link time is the big one @davidbeckingsale, it also makes it easier to do things like filter the tests and use the other advanced features of google test.  It also makes them more parallelizable.

Plus, we can parse out the results of a big google test binary rather easier than a bunch of little ones for inclusion in things like bamboo or appveyor test reports.
